### PR TITLE
[WIP] Add solving/unification step to type inference.

### PR DESCRIPTION
**Pull request by nathanielc**
This is still WIP. Notes about the current progress:

- [ ] Still need to handle function parameters default values. This probably means breaking keys from default values in the semantic graph
- [ ] Handle functions with node bodies instead of just an expression
- [ ] Handle member expressions

The design uses two key types:

Constraint and Substitutable. A Substitutable represents a type expressions that can substitute a constraint. Three types implement the substitutable interface (Type, TypeVar, and Constraint). The solve/unification step substitutes each of the constraints on all other constraints and type variables. Substituting a TypeVar with a constraint that maps a var to a monotype for example returns the monotype directly. The returned monotype is then placed in the solution map for all nodes that had that type variable.

 I still have some questions:

* Does the solve step need to be a fixed point algorithm? I expected it needed to be, but so far it can solve everything with a single n^2 pass through the constraints. Maybe as we add more test cases this will become clear.
* How should we handle polymorphism? Right now the function expression type var gets substituted into a monotype, which means it cannot be a polytype. 
* How can we effectively test this? As the solution map is map of pointers to Types, its hard to write a test that uses pointers.

Other notes:

* I combined the declarations and function_return bits all into the same constraint generator visitor type. This made it more straightforward to follow since there was no longer a logical dependency through side effects between the different visitors.
* Scope nesting is handled by nesting the visitor itself.
* I explored using variants of the Constraint type like a CallConstraint vs EqualityConstraint. That did not work out well because it became ambiguous how Substitutables should substitue those different Constraints. Instead I added more type annotations, for the return type of functions and created a new constraint between a call expression and the return type type variable.

### CODE REVIEW

**Comment by nathanielc** [comment]
> How can we effectively test this? As the solution map is map of pointers to Types, its hard to write a test that uses pointers.

My latest commit addresses this. Basically each test case provides a visitor that hard codes the expected type given a node from the semantic graph.

File: semantic/solver.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	log.Println("constraints", constraints)
+	hasFreeVars := true
+	for hasFreeVars {
+		hasFreeVars = false
+		for i, a := range constraints {
+			for j, b := range constraints {
+				if i == j {
+					continue
+				}
+				constraints[j] = b.Substitute(a).(Constraint)
```
Comment (jlapacik) [inline semantic/solver.go]: What if you have two constraints with the same type variable on the left hand side:

* `tx` = substitutable type expression
* `tx` = a different substitutable type expression

And `tx` does not appear on the right hand side of any constraint? I think there needs to be substitutions on the left hand side of constraints as well.

**Comment by jlapacik** [comment]
@nathanielc To your first question, I don't believe the solve step needs to be a fixed point algorithm. According to Wand, you should be able to solve everything in n^2 time. One piece that I think is missing that will make this possible is substituting a type expression in the left-hand side of a constraint, if said left-hand side is already bound in the substitution.

File: semantic/constraint.go [hunk 2]
```diff
@@ -223,36 +218,41 @@ func (v *ConstraintGenerationVisitor) Visit(node Node) Visitor {
...
+	case *FunctionParam:
+		// maintain scope
+		v.scope.Set(n.Key.Name, n)
+
+		key := v.tenv[n.Key]
+		def, ok := v.tenv[n.Default]
+		if ok {
+			v.addConstraints(Constraint{
+				left:  key,
+				right: def,
```
[the remark this answers is not in the document]
Comment (nathanielc) [reply semantic/constraint.go]: No, if a parameter has a default value, that parameter is not polymorphic.

File: semantic/constraint.go [hunk 3]
```diff
@@ -201,14 +184,26 @@ func (v *ConstraintGenerationVisitor) TypeEnvironment() map[Node]TypeVar {
...
+		// maintain scope
+		v.scope.Set(n.Identifier.Name, n)
 		// Inference Rule: Variable Declaration
 		// ------------------------------------
 		// x = expression
 		//
 		// -> typeof(x) = typeof(expression)
 		// ------------------------------------
-		v.cons = append(v.cons, Constraint{
+		v.addConstraints(Constraint{
```
[the remark this answers is not in the document]
Comment (nathanielc) [reply semantic/constraint.go]: No, I haven't explicitly handled that case. I'll add a note

File: semantic/constraint.go [hunk 4]
```diff
@@ -223,36 +251,41 @@ func (v *ConstraintGenerationVisitor) Visit(node Node) Visitor {
...
-		//
-		// -> typeof(f) = (typeof(a), typeof(b)) => typeof(f(a:1, b:2))
-		// ------------------------------------------------------------
-		operator := v.tenv[n.Callee]
-		operand := make(map[string]Substitutable, len(n.Arguments.Properties))
-		for i, prop := range n.Arguments.Properties {
-			operand[prop.Key.Name] = v.tenv[n.Arguments.Properties[i].Value]
+	case *FunctionParam:
+		// maintain scope
+		v.scope.Set(n.Key.Name, n)
```
[the remark this answers is not in the document]
Comment (nathanielc) [reply semantic/constraint.go]: No, I its not, I'll fix.

**Pull request closed without merging**